### PR TITLE
Add Numba to rolling mean implementation

### LINKED ISSUE #17: Use Numba to improve performance of rolling mean
Here is the current performance of  rolling mean after https://github.com/twosigma/pandas/pull/6

```
       before           after         ratio
     [5c0da7dd]       [e67a634f]
     <master^2>       <feature/generalized_window_operations>
+     1.12±0.03ms         899±10ms   804.92  rolling.Methods.time_rolling('DataFrame', 1000, 'float', 'mean')
+     1.17±0.01ms          927±5ms   792.58  rolling.Methods.time_rolling('DataFrame', 1000, 'int', 'mean')
+     1.10±0.01ms         824±20ms   747.57  rolling.Methods.time_rolling('DataFrame', 10, 'float', 'mean')
+     1.41±0.01ms          901±4ms   636.93  rolling.Methods.time_rolling('Series', 1000, 'float', 'mean')
+     1.43±0.03ms         906±20ms   633.21  rolling.Methods.time_rolling('Series', 1000, 'int', 'mean')
+     1.33±0.07ms         830±20ms   626.19  rolling.Methods.time_rolling('DataFrame', 10, 'int', 'mean')
+     1.43±0.02ms          830±3ms   581.78  rolling.Methods.time_rolling('Series', 10, 'float', 'mean')
+     1.42±0.01ms          820±6ms   577.79  rolling.Methods.time_rolling('Series', 10, 'int', 'mean')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

The reason is the use of pure python for calculating rolling mean instead of using the current cython c extensions. Numba should be able to get us parity with cython performance.

Side note: See this numba issue for a performance consideration when using numba
https://github.com/numba/numba/issues/2166

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #17

### CODE REVIEW

File: pandas/core/window/indexers.py [hunk 1]
```diff
@@ -78,8 +79,9 @@ def get_window_bounds(
 class FixedWindowIndexer(BaseIndexer):
     """Calculate window boundaries that have a fixed window size"""
 
+    @staticmethod
```
Comment (mroeschke) [inline pandas/core/window/indexers.py]: I hope I won't need to follow this pattern.

Currently I am having some struggles getting `jitclass` to work with classes inheriting from `abc.ABC` (as mentioned on gitter)

**Comment by mroeschke** [comment]
Performance after https://github.com/twosigma/pandas/pull/21/commits/baed4bc71540026aed68caa40953f45c0bafaa74

```
       before           after         ratio
     [9c748c69]       [baed4bc7]
     <master>         <feature/numba_performance>
+     1.11±0.02ms          220±3ms   197.77  rolling.Methods.time_rolling('DataFrame', 1000, 'float', 'mean')
+        1.16±0ms          224±4ms   193.20  rolling.Methods.time_rolling('DataFrame', 1000, 'int', 'mean')
+     1.42±0.01ms          226±2ms   159.06  rolling.Methods.time_rolling('Series', 1000, 'float', 'mean')
+     1.49±0.05ms          219±1ms   147.55  rolling.Methods.time_rolling('Series', 1000, 'int', 'mean')
+     1.16±0.02ms        133±0.3ms   115.38  rolling.Methods.time_rolling('DataFrame', 10, 'float', 'mean')
+     1.44±0.02ms          138±3ms    95.87  rolling.Methods.time_rolling('Series', 10, 'int', 'mean')
+      1.45±0.2ms          136±2ms    93.83  rolling.Methods.time_rolling('DataFrame', 10, 'int', 'mean')
+     1.44±0.03ms          133±1ms    92.29  rolling.Methods.time_rolling('Series', 10, 'float', 'mean')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

**Comment by jreback** [comment]
hmm where is performance decrease concentrated?

**Comment by jreback** [comment]
is this running in python mode?

**Comment by mroeschke** [comment]
`jitclass` should run all class methods in nopython mode:

https://numba.pydata.org/numba-doc/dev/user/jitclass.html

This last function is not jitted yet which where the performance bottleneck might be:

https://github.com/twosigma/pandas/pull/21/files#diff-0a4dda541b93b21c24e2acfefd0c08d4R210

**Comment by mroeschke** [comment]
Performance after https://github.com/twosigma/pandas/pull/21/commits/a21a999b268ff1e47b05df3ab69664264ed7165a:

```
       before           after         ratio
     [4a1ae8b0]       [a21a999b]
     <master>         <feature/numba_performance>
+     1.14±0.05ms          103±1ms    90.27  rolling.Methods.time_rolling('DataFrame', 1000, 'float', 'mean')
+     1.21±0.05ms          103±1ms    85.49  rolling.Methods.time_rolling('DataFrame', 1000, 'int', 'mean')
+     1.42±0.04ms        103±0.7ms    72.45  rolling.Methods.time_rolling('Series', 1000, 'float', 'mean')
+     1.49±0.04ms          103±1ms    69.32  rolling.Methods.time_rolling('Series', 1000, 'int', 'mean')
+      1.21±0.3ms       18.9±0.3ms    15.64  rolling.Methods.time_rolling('DataFrame', 10, 'int', 'mean')
+      1.23±0.1ms       18.9±0.8ms    15.45  rolling.Methods.time_rolling('DataFrame', 10, 'float', 'mean')
+     1.46±0.07ms       18.9±0.4ms    12.92  rolling.Methods.time_rolling('Series', 10, 'float', 'mean')
+      1.65±0.1ms       18.9±0.3ms    11.47  rolling.Methods.time_rolling('Series', 10, 'int', 'mean')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

**Comment by jreback** [comment]
any better on benchmarks?

**Comment by mroeschke** [comment]
Adding `nogil=True` and `parallel=True` didn't improve performance. Working on not using `jitclass` so I can utilize `nogil` in other places.

**Comment by mroeschke** [comment]
As of https://github.com/twosigma/pandas/pull/21/commits/735fef1aa06a9eb9df49b655396d959c1ff9f006, I was able to improve performance a little bit, and I also tested two different numba approaches to calculate the rolling mean.

1) One using the current `jitclass` infrastructure.
2) One using a singular `njit` method

The results are below, but overall the singular njit method is still faster than the jitclass set up. The njit method also gets pretty close to the current cython performance.

https://github.com/numba/numba/issues/4055 might explain the differences

Here's the performance with using the jitclasses.
```
       before           after         ratio
     [b638376d]       [735fef1a]
     <master>         <feature/numba_performance>
+      1.14±0.3ms      13.1±0.08ms    11.48  rolling.Methods.time_rolling('DataFrame', 10, 'float', 'mean')
+      1.23±0.2ms       13.7±0.4ms    11.20  rolling.Methods.time_rolling('DataFrame', 10, 'int', 'mean')
+      1.22±0.1ms      13.2±0.04ms    10.85  rolling.Methods.time_rolling('DataFrame', 1000, 'int', 'mean')
+      1.26±0.1ms       13.1±0.1ms    10.42  rolling.Methods.time_rolling('DataFrame', 1000, 'float', 'mean')
+      1.59±0.2ms       14.2±0.4ms     8.89  rolling.Methods.time_rolling('Series', 1000, 'int', 'mean')
+      1.57±0.1ms       13.8±0.3ms     8.75  rolling.Methods.time_rolling('Series', 10, 'int', 'mean')
+      1.63±0.3ms       14.1±0.4ms     8.68  rolling.Methods.time_rolling('Series', 1000, 'float', 'mean')
+      1.83±0.3ms       14.0±0.5ms     7.68  rolling.Methods.time_rolling('Series', 10, 'float', 'mean')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

Here's the performance using a single njit method

```
       before           after         ratio
     [b638376d]       [efe42346]
     <master>         <feature/numba_performance>
+      1.22±0.1ms       3.41±0.3ms     2.80  rolling.Methods.time_rolling('DataFrame', 10, 'int', 'mean')
+      1.19±0.1ms       3.22±0.1ms     2.69  rolling.Methods.time_rolling('DataFrame', 1000, 'float', 'mean')
+      1.30±0.1ms       3.46±0.2ms     2.66  rolling.Methods.time_rolling('DataFrame', 1000, 'int', 'mean')
+     1.42±0.05ms       3.70±0.2ms     2.60  rolling.Methods.time_rolling('Series', 1000, 'float', 'mean')
+     1.50±0.09ms       3.72±0.1ms     2.48  rolling.Methods.time_rolling('Series', 10, 'float', 'mean')
+     1.61±0.07ms       3.73±0.2ms     2.32  rolling.Methods.time_rolling('Series', 1000, 'int', 'mean')
+      1.47±0.3ms      3.11±0.04ms     2.11  rolling.Methods.time_rolling('DataFrame', 10, 'float', 'mean')
+      2.07±0.3ms       3.79±0.1ms     1.83  rolling.Methods.time_rolling('Series', 10, 'int', 'mean')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

File: pandas/tests/window/test_rolling.py [hunk 2]
```diff
@@ -327,6 +327,7 @@ def test_rolling_axis_count(self, axis_frame):
         result = df.rolling(2, axis=axis_frame).count()
         tm.assert_frame_equal(result, expected)
 
+    @pytest.mark.xfail(reason="Numba doesn't support readonly arrays?")
```
Comment (mroeschke) [inline pandas/tests/window/test_rolling.py]: https://github.com/numba/numba/issues/4511

File: pandas/core/window/aggregators.py [hunk 3]
```diff
@@ -160,32 +162,83 @@ def combine(self, other) -> None:
...
+    # aggregator = kernel_class().make_aggregator(values, minimum_periods)
     result = np.empty(len(begin))
     for i, (start, stop) in enumerate(zip(begin, end)):
         result[i] = aggregator.query(start, stop)
     return result
 
 
-rolling_mean = partial(rolling_aggregation, kernel_class=Mean)
+@numba.njit(aggregation_signature, nogil=True, parallel=True)
+def rolling_mean_method(
```
Comment (mroeschke) [inline pandas/core/window/aggregators.py]: Need to move this into a separate file.

1) One file should have the jitclass implementations (kernels and aggregators)
2) One file will have the njit implementations (current cython "style")

**Pull request merged**